sdb/commands/stacks.py

```diff
@@ -216,86 +248,120 @@ def find_module_memory_segment(mod_name: str) -> Tuple[int, int]:
+            if not self.args.tstate in task_states:
+                raise sdb.CommandError(
+                    self.name,
+                    "'{:s}' is not a valid task state (acceptable states: {:s})"
+                    .format(self.args.tstate, ", ".join(task_states)))
 
-        mod_start, mod_end = -1, -1
         if self.args.module:
-            mod_start, mod_size = Stacks.find_module_memory_segment(
+            if Stacks.find_module_memory_segment(self.args.module)[0] == -1:
+                raise sdb.CommandError(
+                    self.name,
+                    "module '{:s}' doesn't exist or isn't currently loaded".
+                    format(self.args.module))
+            self.mod_end, self.mod_start = Stacks.find_mod_offset(
                 self.args.module)
-            assert mod_start != -1
-            mod_end = mod_start + mod_size
 
+    def print_header(self):
         header = "{:<18} {:<16s}".format("TASK_STRUCT", "STATE")
         if not self.args.all:
             header += " {:>6s}".format("COUNT")
         print(header)
         print("=" * 42)
 
+    #
+    # Return a boolean indicating if this stack matches the state, function or module
+    # arguments (if any) and a string representation of the stack.
+    #
+    def match_stack(self, stack_key: Tuple[Any], tasks: List[drgn.Object],
+                    print_cb: Callable[[str], None]):
+        match, stacktrace_info = True, ""
+
+        task_state = stack_key[0]
+        if self.args.tstate and self.args.tstate != task_state:
+            return False, stacktrace_info
+
+        if self.args.all:
+            for task in tasks:
+                stacktrace_info += "{:<18s} {:<16s}\n".format(
+                    hex(task.value_()), task_state)
+        else:
+            stacktrace_info += "{:<18s} {:<16s} {:6d}\n".format(
+                hex(tasks[0].value_()), task_state, len(tasks))
+        mod_match, func_match = False, False
         #
-        # We inspect and group the tasks by recording their state and
-        # stack frames once in the following loop. We do this because
-        # on live systems state can change under us, thus running
-        # something like sdb.get_prog().stack_trace(task) twice (once for
-        # grouping and once for printing) could yield different stack
-        # traces resulting into misleading output.
+        # Note on the type-check being ignored:
+        # The original `stack_key` type is a list where the first
+        # element is a string and the rest of them are integers
+        # but this is not easily expressed in mypy, thus we ignore
+        # the assignment error below.
         #
+        frame_pcs: Tuple[int] = stack_key[1:]  # type: ignore[assignment]
+        for frame_pc in frame_pcs:
+            if self.args.module and self.mod_start <= frame_pc < self.mod_end:
+                mod_match = True
+
+            try:
+                sym = sdb.get_symbol(frame_pc)
```

Comment:
From our previous discussion, it sounded like `sdb.get_symbol()` was slow (at least when called on every frame of every stack).  If we don't care about printing, and aren't filtering on function, we could avoid this call.  Even if we are filtering on function, we could avoid this call by saving the function's offset and length, like we do for the module.

Once we've avoided symbol lookup when we aren't printing, it would be performant to separate filtering from printing, which I think would be a nice refactoring of this code.  It would probably also be performant to do filtering before aggregation, which would more cleanly separate finding/filtering from aggregating/printing.

For an example of filtering and printing separated, see dbuf.py's no_input, pretty_print, and argfilter methods.

Reply:
I updated the PR to incorporate these suggestions (with the caveat that we still need to filter in the printing case).

Let me know what you think!